datalad_next/itertools/__init__.py

```diff
@@ -1,13 +1,270 @@
+   by ``align_pattern`` ends with a prefix of the pattern (unless it is the last
+   chunk that the underlying iterable yield).
+
+Note: the pattern is taken verbatim, and not interpreted as a regular expression.
+
+.. code-block:: python
+
+    >>> from datalad_next.itertools import align_pattern
+    >>> tuple(align_pattern([b'abcd', b'e', b'fghi'], pattern=b'def'))
+    (b'abcdefghi',)
+    >>> # The pattern can be present multiple times in a yielded chunk
+    >>> tuple(align_pattern([b'abcd', b'e', b'fdefghi'], pattern=b'def'))
+    (b'abcdefdefghi',)
+
+Use this function if you want to locate a pattern in an input stream. It allows
+to use a simple ``in``-check to determine whether the pattern is present in the
+yielded result chunks.
+
+
+Usage of ``decode_bytes``
+-------------------------
+
+``decode_bytes`` allows its caller to decode ``bytes``- or
+``bytearray``-objects  into ``str``-objects, using the specified encoding
+(default encoding: 'utf-8').
+
+Whenever there is the need to decode bytes that have been split into
+chunks of random sizes, ``decode_bytes`` should be used. ``decode_bytes``
+takes care to properly decode multi-byte encodings
+that are spread over multiple byte chunks. For example, the utf-8
+encoding for ö is ``b'\\xc3\\xb6'``. If the encoding is split in the
+middle because a chunk ends with ``b'\\xc3'`` and the
+next chunk starts with ``b'\\xb6'``, a naive decoding approach like the
+following would fail:
+
+.. code-block:: python
+
+    >>> [chunk.decode() for chunk in [b'\\xc3', b'\\xb6']]
+    Traceback (most recent call last):
+      File "<stdin>", line 1, in <module>
+      File "<stdin>", line 1, in <listcomp>
+    UnicodeDecodeError: 'utf-8' codec can't decode byte 0xc3 in position 0: unexpected end of data
+
+If ``decode_bytes`` is used, the chunks are decoded correctly:
+
+.. code-block:: python
+
+    >>> from datalad_next.itertools import decode_bytes
+    >>> tuple(decode_bytes([b'\\xc3', b'\\xb6']))
+    ('ö',)
+
+``decode_bytes`` will join input chunks only if that becomes necessary to
+properly decode bytes strings.
+
+.. code-block:: python
+
+    >>> from datalad_next.itertools import decode_bytes
+    >>> tuple(decode_bytes([b'\\xc3', b'\\xb6', b'a']))
+    ('ö', 'a')
+
+If ``backslash_replace`` is ``True`` (the default), undecodable bytes will be
+replaced with a backslash-substitution. If ``backslash_replace`` is ``False``,
+undecodable bytes will raise a ``UnicodeDecodeError``:
+
+.. code-block:: python
+
+    >>> tuple(decode_bytes([b'\\xc3']))
+    ('\\xc3',)
+    >>> tuple(decode_bytes([b'\\xc3'], backslash_replace=False))
+    Traceback (most recent call last):
+        ...
+    UnicodeDecodeError: 'utf-8' codec can't decode byte 0xc3 in position 1: invalid continuation byte
+
+Note that backslash-replacement of undecodable bygtes does not result in an
+unambiguous mapping, because, for example, ``b'\\xc3'`` might
+be present in the input chunks.
```

Comment:
It is unclear to me what that means.

Reply:
Ok, so:

```
Note that backslash-replacement of undecodable bytes does not result in an
unambiguous mapping, because, for example, ``b'\\xc3'`` might
be present in the input chunks.
```

should become

```
Backslash-replacement of undecodable bytes is an ambiguous mapping, because, for example, ``b'\\xc3'`` can already be present in the input.
```

Reply:
Thanks @mih Much better!

Reply:
Fixed in https://github.com/datalad/datalad-next/pull/548